### PR TITLE
Add support for content-type-based decoding of request body

### PR DESCRIPTION
**Pull request by sergeykolbasov**
https://github.com/finagle/finch/issues/958

- Add new abstraction `AdaptableDecode` that is used only for `body` and all the related endpoints. This abstraction enables coproduct support for `Content-Type`:
```scala
body[A, Application.Json :+: Application.Csv :+: CNil]
```
Backward-compatibility is there with a very small overhead of one more instance per call.

- Add specs and related documentation

### CODE REVIEW

**Comment by sergeykolbasov** [comment]
@vkostyukov I'm not sure about calling headers (Content-Type) each time, but taking in account that we already consume the whole body of a request, I think it should be relatively fast.

File: core/src/main/scala/io/finch/Decode.scala [hunk 1]
```diff
@@ -40,3 +41,36 @@ object Decode {
    */
   @inline def apply[A, CT <: String](implicit d: Aux[A, CT]): Aux[A, CT] = d
 }
+
+trait AdaptableDecode[A, CT] {
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Decode.scala]: Maybe we can move this under `Decode` companion object, by analogy with `ToResponse.Negotiable`?

Also, what do you think about renaming this to `Decode.Dispatchable`?

File: core/src/main/scala/io/finch/Decode.scala [hunk 2]
```diff
@@ -40,3 +41,36 @@ object Decode {
...
+}
+
+object AdaptableDecode {
+
+  def instance[A, CT](fn: Option[String] => Decode.Aux[A, CT]):  AdaptableDecode[A, CT] = new AdaptableDecode[A, CT] {
+    def apply(contentType: Option[String]): Decode.Aux[A, CT] = fn(contentType)
+  }
+
+  implicit def mkCnil[A, CTH <: String](implicit decode: Decode.Aux[A, CTH]): AdaptableDecode[A, CTH :+: CNil] =
+    instance(_ => decode.asInstanceOf[Decode.Aux[A, CTH :+: CNil]])
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Decode.scala]: I like that this preserves the behavior, but maybe in future, we'd need to change this and make sure to check if the given content-type matches.

Comment (vkostyukov) [reply core/src/main/scala/io/finch/Decode.scala]: Same thing about `mkSingle`.

File: core/src/main/scala/io/finch/endpoint/body.scala [hunk 3]
```diff
@@ -125,15 +125,15 @@ private[finch] trait Bodies {
    * (non-streamed) requests.
    */
   def bodyOption[A, CT <: String](implicit
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/endpoint/body.scala]: We should probably remove `<: String` here too.

File: core/src/main/scala/io/finch/Decode.scala (hunk 1 above)
Comment (vkostyukov) [reply core/src/main/scala/io/finch/Decode.scala]: Thinking out loud. I sort of like that we had `CT <: String` requirement on the `Decode` type-class. Perhaps, we can work it around and keep it in place for the sake of symmetry with `Encode`?

One thing we can do is to have a new type-class wrapping `Decode` (not returning it). Similar to how `ToResponse` and `Encode` relation is modeled.

Quick sketch. Let me know if you like it.

```scala
object Decode {
  
  trait Dispatchable[A, CT] {
    def apply(ct: Option[String], b: Buf, cs: Charset): Try[A]
  }
}
```

Comment (sergeykolbasov) [reply core/src/main/scala/io/finch/Decode.scala]: Sounds good

**Comment by sergeykolbasov** [comment]
@vkostyukov there is a good article about compiler profiling:
https://www.scala-lang.org/blog/2018/06/04/scalac-profiling.html

Probably, we could set up something on Travis CI, but it should be different issue already

File: core/src/main/scala/io/finch/endpoint/body.scala [hunk 4]
```diff
@@ -55,33 +55,33 @@ private object FullBody {
   }
 }
 
-private abstract class Body[A, B, CT <: String](
-  d: Decode.Aux[A, CT],
-  ct: ClassTag[A]
+private abstract class Body[A, B, CT](
+                                       ad: Decode.Dispatchable[A, CT],
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/endpoint/body.scala]: What happened to formatting here?

File: core/src/main/scala/io/finch/Decode.scala [hunk 5]
```diff
@@ -39,4 +40,40 @@ object Decode {
...
+      instance((contentType, b, cs) => {
+        if (contentType.exists(ct => ct.equalsIgnoreCase(witness.value))) {
+          decode(b, cs)
+        } else {
+          tail(contentType, b, cs)
+        }
+      })
+
+    implicit def mkSingle[A, CT <: String](implicit
+                                           decode: Decode.Aux[A, CT]
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Decode.scala]: Formatting seems off.

File: core/src/main/scala/io/finch/Decode.scala [hunk 6]
```diff
@@ -39,4 +40,40 @@ object Decode {
...
+
+    def instance[A, CT](fn: (Option[String], Buf, Charset) => Try[A]):  Dispatchable[A, CT] = new Dispatchable[A, CT] {
+      def apply(contentType: Option[String], b: Buf, cs: Charset): Try[A] = fn(contentType, b, cs)
+    }
+
+    implicit def mkCnil[A, CTH <: String](implicit decode: Decode.Aux[A, CTH]): Dispatchable[A, CTH :+: CNil] =
+      instance((_, b, cs) => decode(b, cs))
+
+    implicit def mkCons[A, CTH <: String, CTT <: Coproduct](implicit
+                                                            decode: Decode.Aux[A, CTH],
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Decode.scala]: Same q about formatting.

**Comment by rpless** [comment]
Selecting the last decoder when the coproduct doesn't contain a Content-Type is interesting. I'm not sure this is the behavior I would expect. Part of me feels like the endpoint should not get matched in that case. Reading through the [spec](https://tools.ietf.org/html/rfc7231#section-3.1.1.5) though seems to indicate that this behavior is fine. I think we'll need to make sure we document this clearly in the User Guide.
Other than that, :+1: from me.

**Comment by vkostyukov** [comment]
Yeah, I somewhat agree with @rpless that perhaps we should adjust this behavior in the future and either return 404 (not matched) or 400 (not parsed). We, however, may need to keep the current logic in place and do not inspect content-type before applying a decoder when there is only one decoder in the list (when we reach the last).  I filed #967 so we can discuss this.

**Pull request merged**